tsnake/initialize.py

```diff
@@ -0,0 +1,543 @@
+        edge_pixels = self._find_edge_pixels()
+        
+        step = int(np.floor(len(edge_pixels) / N))
+        step = max(step, 1)
+        if verbose:
+            print('Initializing a Node at every %d-th boundary-pixel' % step)
+        
+        # Order the nodes so that they connect to closest nodes
+        ordered_nodes = self.order_snake_nodes(edge_pixels)
+        
+        # Pull out every step-th pixel for initializion as a Node
+        nodes = []
+        for node_num in range(0, len(ordered_nodes), step):
+            r, c = ordered_nodes[node_num]
+            new_node = Node(r, c)
+            nodes.append(new_node)
+            
+        if verbose:
+            print('Total of %d T-snake nodes were initialized' % len(nodes))
+        
+        force_grid = self.compute_force_grid(sigma, c)
+        intensity_grid = self.raw_image_portion  # grayscale image
+        
+        snake = TSnake(
+            nodes, force_grid, intensity_grid, a, b, gamma, dt
+        )
+        self._initial_tsnake = snake
+        return snake
+    
+    def order_snake_nodes(self, node_locs):
+        """
+        TODO: doc
+        """
+        num_nodes = len(node_locs)
+        node_locs = np.array(node_locs).reshape(num_nodes, 2)
+        
+        # Pull out arbitrary starting node
+        loc = node_locs[0]
+        node_locs = node_locs[1:]
+        
+        ordered_nodes = [loc]
+        while len(ordered_nodes) < num_nodes:
+            # Find the node which is closest to the last node we processed
+            last = ordered_nodes[-1].reshape(1, 2)
+            # dists = distance from each remaining node to current node
+            dists = cdist(last, node_locs).reshape(-1, )
+            
+            closest = dists.argmin()
+            ordered_nodes.append(node_locs[closest])
+            node_locs = np.delete(node_locs, obj=closest, axis=0)
+        
+        return ordered_nodes
+    
+    def compute_force_grid(self, sigma, c):
+        """
+        NOTE: can we manipulate the force grid to make sure the snake
+              doesn't leave the user-masked area?
+        
+        Args:
+        ======================
+        (float) sigma: 
+        * The hyperparameter sigma from Equation (A.4).
+        
+        (float) c:
+        * The hyperparameter c from Equation (A.4).
+        ======================
+        Returns:
+        ======================
+        A np.array of shape (n, m, 2) containing the computed values of 
+        Equation (7) at each pixel in the image.
+        ======================
+        """
+        # Apply Gaussian smoothing
+        out = gaussian_filter(self.raw_image_portion, sigma=sigma)
+        
+        # Take the gradient
+        x_grad, y_grad = np.gradient(out)
+        out = np.zeros((x_grad.shape[0], x_grad.shape[1], 2))
+        out[:, :, 0] = x_grad
+        out[:, :, 1] = y_grad
+        
+        # Take the magnitude (??)
+        out = np.abs(out)
+        
```

Comment:
So, I reread the paper and now understand how to get the force grid. After taking the gradient of the filtered image, we need to take the magnitude of every pixel vector and multiply by -c. Then, we take the gradient of that and return the numpy array of shape (n,m,2). 

Suggested fix:
#find potential grid which is a scalar for every pixel
out = np.sqrt(np.square(x_grad) + np.square(y_grad))
out = out * -c
#now take gradient of potential 
x_grad, y_grad = np.gradient(out)
out = np.zeros((x_grad.shape[0], x_grad.shape[1], 2))
out[:, :, 0] = x_grad
out[:, :, 1] = y_grad
#force parameter as defined in the paper, equation (7)
out = p * out

Reply:
Fixed